tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1173,191 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+ * encountered, the server either chooses that cipher suite indicated by that index or, if no cipher suite 
+ * has been found, continues iterating through its list. This function checks that the potential cipher
+ * suite match is valid and can be used for the connection.
+ * 
+ */
+static S2N_RESULT s2n_get_mutually_supported_cipher_index(struct s2n_connection* conn,
+                                                          const uint8_t* wire,
+                                                          uint32_t count,
+                                                          uint32_t cipher_suite_len,
+                                                          uint32_t* server_index) {
+    RESULT_ENSURE_REF(wire);
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(server_index);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    RESULT_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    /** 
+     * Tracks mutually supported cipher suite in the client index. This is needed to track tie-breakers
+     * in an equal-preference group.
+     */
+    int negotiated_client_index = count;
+    /* Same as negotiated_client_index, but is instead a server index. */
+    int negotiated_server_index = -1;
+    /** 
+     * The highest protocol version match index (also a server index). This is a fall back index if
+     * no other mutually-supported cipher suite is identified.
+     */
+    int highest_vers_match_index = -1;
+
+    for (uint32_t i = 0; i < cipher_preferences->count; i++) {
+        const struct s2n_cipher_suite* server_cipher = cipher_preferences->suites[i];
+        RESULT_ENSURE_REF(server_cipher);
+
+        /* Check if the cipher suite is an equal-preference grouping delimiter */
+        if (server_cipher == &s2n_equal_preference_group_start) {
+            in_group = true;
+            continue;
+        } 
+        if (server_cipher == &s2n_equal_preference_group_end) {
+            in_group = false;
+            /* Exiting a group and a negotiated cipher has already been found. */
+            if (negotiated_server_index != -1) {
+                *server_index = negotiated_server_index;
+                return S2N_RESULT_OK;
+            }
+            continue;
+        }
+
+        /* Cipher suite is NOT a delimiter */
+        int client_index = s2n_wire_index_of_cipher_suite(server_cipher->iana_value, wire, count, cipher_suite_len);
+
+        /* Client does not support this cipher. Skip. */
+        if (client_index < 0) {
+            continue;
+        }
+
+        /* Found a potential match. Validate that the cipher suite is suitable for this connection. */
+        struct s2n_cipher_suite* match = security_policy->cipher_preferences->suites[i];
```

Comment:
Why do you need this match variable? You already have a pointer to the cipher suite with server_cipher right?

Reply:
Will fix